flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java

```diff
@@ -0,0 +1,272 @@
+                                                Row.of(2, "HeHe", "Hi5"),
+                                                Row.of(1, "HiHi", "Hi6"),
+                                                Row.of(1, "HiHi", "Hi8"),
+                                                Row.of(null, "HeHe", "Hi9"))
+                                        .consumedAfterRestore(Row.of(2, "PostRestoreRight", "Hi5"))
+                                        .build())
+                        .runSql(
+                                "insert into MySink "
+                                        + "SELECT t2.a, t2.c, t1.c\n"
+                                        + "FROM (\n"
+                                        + " SELECT if(a = 3, cast(null as int), a) as a, b, c FROM T1\n"
+                                        + ") as t1\n"
+                                        + "JOIN (\n"
+                                        + " SELECT if(a = 3, cast(null as int), a) as a, b, c FROM T2\n"
+                                        + ") as t2\n"
+                                        + "ON \n"
+                                        + "  ((t1.a is null AND t2.a is null) OR\n"
+                                        + "  (t1.a = t2.a))\n"
+                                        + "  AND t1.b > t2.b")
+                        .build();
+
+        JOIN =
+                TableTestProgram.of("join", "test join")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a3 varchar", "b4 varchar")
+                                        .consumedBeforeRestore(
+                                                Row.of("Hi", "Hallo"),
+                                                Row.of("Hello", "Hallo Welt"),
+                                                Row.of("Hello world", "Hallo Welt"),
+                                                Row.of("Hi", "Hallo Welt wie gehts"))
+                                        .consumedAfterRestore(
+                                                Row.of("Hello there", "Hallo Welt wie"))
+                                        .build())
+                        .runSql("insert into MySink " + "SELECT a3, b4 FROM A, B WHERE a2 = b2")
+                        .build();
+
+        INNER_JOIN =
+                TableTestProgram.of("inner join", "test inner join")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a1 int", "b1 int")
+                                        .consumedBeforeRestore(
+                                                Row.of(1, 1),
```

Comment:
could we use outpuit columns which would make it easier to tell which rows actually got joined? e.g.  `a3` and `b4`

Reply:
The next test looks really good!

Reply:
Thanks!